retrieval/target.go

```diff
@@ -363,12 +368,32 @@ func (t *Target) scrape(sampleAppender storage.SampleAppender) (err error) {
 
 	for samples := range t.ingestedSamples {
 		for _, s := range samples {
-			s.Metric.MergeFromLabelSet(baseLabels, clientmodel.ExporterLabelPrefix)
+			if t.honorLabels {
+				// Merge the metric with the baseLabels for labels not already set in the
+				// metric. This also considers labels explicitly set to the empty string.
+				for ln, lv := range baseLabels {
+					if v, ok := s.Metric[ln]; !ok {
+						s.Metric[ln] = lv
+					} else if v == "" {
+						// Delete empty labels before ingestion.
+						delete(s.Metric, ln)
```

Comment:
This deletes empty labels only for the special case where the label name is in the base labels, and `honor_labels` is set. Should we have a general place after this if-else where we go through *all* resulting labels and remove all empty ones before ingestion? If the target exports `foo=""`, that should disappear as well, right?